dll/cpl/intl/languages.c

```diff
@@ -2,30 +2,45 @@
 
 #include <shellapi.h>
 #include <shlobj.h>
+#include <shlwapi.h>
 #include <strsafe.h>
 
-/* Where is the uninstaller of "ReactOS JPN Package"? */
-BOOL GetJapaneseUninstallPath(HWND hwnd, LPWSTR pszPath, SIZE_T cchPath)
+/* What is the uninstallation command line of "ReactOS JPN Package"? */
+BOOL GetJapaneseUninstallCmdLine(HWND hwnd, LPWSTR pszCmdLine, SIZE_T cchCmdLine)
 {
-    SHGetSpecialFolderPathW(hwnd, pszPath, CSIDL_PROGRAM_FILES, FALSE);
-    StringCchCatW(pszPath, cchPath, L"\\ReactOS JPN Package\\unins000.exe");
-    if (GetFileAttributesW(pszPath) != INVALID_FILE_ATTRIBUTES)
-        return TRUE;
-
-    SHGetSpecialFolderPathW(hwnd, pszPath, CSIDL_PROGRAM_FILESX86, FALSE);
-    StringCchCatW(pszPath, cchPath, L"\\ReactOS JPN Package\\unins000.exe");
-    if (GetFileAttributesW(pszPath) != INVALID_FILE_ATTRIBUTES)
-        return TRUE;
+    HKEY hKey;
+    LONG error;
+    DWORD dwSize;
+
+    pszCmdLine[0] = UNICODE_NULL;
+
+    error = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                          L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall\\"
+                          L"{80F03D6E-0549-4202-BE81-FF583F56A7A8}_is1",
+                          0,
+                          KEY_READ,
+                          &hKey);
+    if (error != ERROR_SUCCESS)
+        return FALSE;
+
+    dwSize = cchCmdLine * sizeof(WCHAR);
+    error = RegQueryValueExW(hKey, L"UninstallString", NULL, NULL, (LPBYTE)pszCmdLine, &dwSize);
+    if (error != ERROR_SUCCESS)
+    {
+        RegCloseKey(hKey);
+        return FALSE;
+    }
 
-    pszPath[0] = UNICODE_NULL;
-    return FALSE;
+    pszCmdLine[cchCmdLine - 1] = UNICODE_NULL;
```

Comment:
Shouldn't it be instead terminated at the end of the retrieved string?

Reply:
`RegQueryValueEx.REG_SZ` does it. The other cases doesn't terminate the buffer.